### PR TITLE
Reduce amount of allocated MissingMetadataTypeSymbol symbols during various lookup operations.

### LINKED ISSUE #64142: Performance: Improve `GetTypeByMetadataName` fail path
I'm in a case where the fail-path of `GetTypeByMetadataName` is hit **a lot** causing lots of allocations here:

https://github.com/dotnet/roslyn/blob/cd5e6895d3dc8dfad270505b37d28be70077a645/src/Compilers/CSharp/Portable/Symbols/NonMissingModuleSymbol.cs#L186-L203

image

Note that this allocation is thrown away because `IsAcceptableMatchForGetTypeByMetadataName` will be false:

https://github.com/dotnet/roslyn/blob/cd5e6895d3dc8dfad270505b37d28be70077a645/src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs#L975-L981

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #64142.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by Youssef1313** [comment]
Thank you @AlekseyTs. Sorry that I haven't had a chance yet to do it in https://github.com/dotnet/roslyn/pull/64176. I'm going to close my PR in favor of this one.

**Comment by Youssef1313** [comment]
> Related to #64142.

I'm curious why this is not "fixes/closes". Is there something remaining?

**Comment by AlekseyTs** [comment]
> Is there something remaining?

Need to do the same for VB

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb [hunk 1]
```diff
@@ -503,7 +503,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End If
 
             Dim emittedName = MetadataTypeName.FromFullName(metadataName, useCLSCompliantNameArityEncoding:=True)
-            Return Me.ContainingModule.LookupTopLevelMetadataType(emittedName)
+            Dim result As NamedTypeSymbol = Me.ContainingModule.LookupTopLevelMetadataType(emittedName)
+            Debug.Assert(If(Not result?.IsErrorType(), True))
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Binding/Binder.vb:507]: Is there a reason we can't just do `result?.IsErrorType() IsNot True`? I find the coalescing operator hard to read here.

**Review by 333fred — COMMENTED** [review]
> Need to do the same for VB

What is missing from the VB implementation in this PR?

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/Binder.vb:507]: An `Is` or `IsNot` expression is used to do a reference equality comparison.

Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Binding/Binder.vb:507]: Is there a non-coalescing version of this that we can use then?

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/Binder.vb:507]: > Is there a non-coalescing version of this that we can use then?

The other forms I considered:
- ```Not ((result?.IsErrorType()).GetValueOrDefault())```
- ```result is Nothing OrElse Not result.IsErrorType()```

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb (hunk 1 above)
Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Binding/Binder.vb:507]: Alright. I don't think any of those are enough of an improvement to insist on it.

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
>>Need to do the same for VB

> What is missing from the VB implementation in this PR?

The third commit covered that

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Comment by 333fred** [comment]
> > > Need to do the same for VB
> 
> > What is missing from the VB implementation in this PR?
> 
> The third commit covered that

In that case, consider updating the description to be `Fixes #64142`.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**